### PR TITLE
Introduce monero-wallet crate

### PR DESCRIPTION
**Pull request by thomaseizinger**
This PR:

1. ~Introduce a crate for the epee binary serialization as a serde format~: Released here: https://github.com/comit-network/monero-epee-bin-serde
2. Extends the MoneroRPC client with two binary calls
3. Introduces a `monero-wallet` crate that for now just provides functionality for choosing random key offsets. Together with the the ability to produce bulletproofs and ring signatures, this should be enough for signing Monero transactions locally.

(1) and (2) are a prerequisite for (3).

### CODE REVIEW

File: swap/src/bin/swap.rs [hunk 1]
```diff
@@ -103,17 +103,16 @@ async fn main() -> Result<()> {
...
+                bitcoin_wallet,
                 Arc::new(monero_wallet),
                 env_config,
-                event_loop_handle,
+                swap_id,
                 receive_monero_address,
-            )
-            .with_init_params(send_bitcoin)
-            .build()?;
+            );
```
Comment (thomaseizinger) [inline swap/src/bin/swap.rs:115]: TODO: Keep the order of the parameters the same to reduce noise.

File: monero-epee-bin-serde/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+use crate::de::Deserializer;
+use crate::ser::Serializer;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::io::Read;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+const MAGIC: &[u8] = b"\x01\x11\x01\x01\x01\x01\x02\x01\x01";
```
Comment (rishflab) [inline monero-epee-bin-serde/src/lib.rs:16]: O.o What does this do? Maybe a comment would be good so you don't have to read `from_bytes`?

File: monero-rpc/src/monerod.rs [hunk 3]
```diff
@@ -1,46 +1,92 @@
...
+use serde::{de::DeserializeOwned, Deserialize, Serialize, Serializer};
 
 #[jsonrpc_client::api(version = "2.0")]
 pub trait MonerodRpc {
     async fn generateblocks(&self, amount_of_blocks: u32, wallet_address: String)
         -> GenerateBlocks;
     async fn get_block_header_by_height(&self, height: u32) -> BlockHeader;
     async fn get_block_count(&self) -> BlockCount;
+    async fn get_block(&self, height: u32) -> GetBlockResponse;
+    async fn on_get_block_hash(&self, height: [u32; 1]) -> String;
```
Comment (rishflab) [inline monero-rpc/src/monerod.rs:12]: I cant see this being used anywhere, What does it do?

File: monero-rpc/src/monerod.rs [hunk 4]
```diff
@@ -1,46 +1,92 @@
...
+use monero::{cryptonote::hash::Hash, util::ringct, PublicKey};
+use serde::{de::DeserializeOwned, Deserialize, Serialize, Serializer};
 
 #[jsonrpc_client::api(version = "2.0")]
 pub trait MonerodRpc {
     async fn generateblocks(&self, amount_of_blocks: u32, wallet_address: String)
         -> GenerateBlocks;
     async fn get_block_header_by_height(&self, height: u32) -> BlockHeader;
     async fn get_block_count(&self) -> BlockCount;
+    async fn get_block(&self, height: u32) -> GetBlockResponse;
```
Comment (rishflab) [inline monero-rpc/src/monerod.rs:11]: I cant see this being used anywhere?

File: monero-wallet/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            .unwrap_or_else(|| block.blob.miner_tx.hash());
+
+        let indices = self.client.get_o_indexes(tx_hash).await?;
+
+        let last_index = indices
+            .o_indexes
+            .into_iter()
+            .max()
+            .context("Expected at least one output index")?;
+        let oldest_index = last_index - (last_index / 100) * 40; // oldest index must be within last 40% TODO: CONFIRM THIS
```
Comment (rishflab) [inline monero-wallet/src/lib.rs:39]: Is this a protocol constraint?

File: monero-rpc/src/monerod.rs (hunk 3 above)
Comment (thomaseizinger) [reply monero-rpc/src/monerod.rs:12]: You are right, turned out to not be needed.

File: monero-rpc/src/monerod.rs (hunk 4 above)
Comment (thomaseizinger) [reply monero-rpc/src/monerod.rs:11]: It is used in `monero-wallet`.

File: monero-wallet/src/lib.rs (hunk 5 above)
Comment (thomaseizinger) [reply monero-wallet/src/lib.rs:39]: I am not fully sure yet. It is blindly copied based on the first caveat here: https://github.com/kayabaNerve/cryptonote-library#caveats

File: monero-rpc/src/monerod.rs [hunk 6]
```diff
@@ -1,46 +1,91 @@
...
+    fn new(host: String, port: u16) -> Result<Self> {
         Ok(Self {
             inner: reqwest::ClientBuilder::new()
                 .connection_verbose(true)
                 .build()?,
-            base_url: format!("http://127.0.0.1:{}/json_rpc", port)
+            base_url: format!("http://{}:{}/json_rpc", host, port)
                 .parse()
                 .context("url is well formed")?,
+            get_o_indexes_bin_url: format!("http://{}:{}/get_o_indexes.bin", host, port)
```
Comment (luckysori) [inline monero-rpc/src/monerod.rs:37]: :open_mouth: So it's a different base URL for these particular calls.

**Review by luckysori — APPROVED** [review]
Amazing work.

File: monero-rpc/src/monerod.rs (hunk 6 above)
Comment (thomaseizinger) [reply monero-rpc/src/monerod.rs:37]: Yes, they are in fact "normal" HTTP calls (except that they don't care about the method you pass in) with a binary payload.

**Pull request merged**